### DISCUSSION TITLE
Ability to hide specific modules in the module browser

### DISCUSSION THREAD

**Discussion by brylie** [DISTRHO/Cardinal#393] [Ideas]
I appreciate the care that has gone into curating and maintaining the module collection that comes with Cardinal. However, for various reasons it would also be nice to selective exclude specific modules from the module browser. While adding modules to my favorites list is one approach it could also be useful to exclude modules from the list so I can explore modules with the “favorites” filter off.

 The 1.x Module Browser module had a “hide” feature to exclude specific modules while browsing. Similarly, the Rack 2 library lets users uninstall specific modules of their choosing.

I am hoping to find a way to have some ability to manually filter or curate the modules shipped with Cardinal in addition to the “favorites” and broader “manufacturer” and “tags” filters. Is there some configuration file or other approach that would allow me to specify a module exclusion list?

Thanks for your consideration,
Brylie (BCO)

**Comment by falkTX** [comment]
As mentioned in #391 this is unsupported.

Cardinal follows Rack for most things, and considering Rack has its own way to allow this that is not possible in Cardinal, such a feature request would likely never happen on Cardinal side.

At the moment Cardinal has zero-configuration from one instance to another.

Even the favourites are per-plugin, per-session. As in, they are saved in the host data/project and never kept across new runs.
If a sorta module blacklist/denylist would even be added to hide things, it would also be per-session, which reduces its usefulness.

I won't do this, don't need it.
If you really want to see something for it, begin by contributing some code yourself.

**Comment by brylie** [comment]
> If you really want to see something for it, begin by contributing some code yourself.

I'm simply trying to get a better understanding of the idea. I may be able to contribute some code to make it happen if I can get a better understanding. In that sense, opening a discussion is an excellent first step.

It is somewhat hurtful to receive responses like "code it yourself" or patronizing rephrases of the original question like "Use the filter functions?" when posting a genuine question in good faith. 

To encourage positive community interactions, perhaps the Cardinal project could adopt the Contributor Covenant, which includes a few guidelines for promoting to a positive community environment, such as the following 

- Demonstrating empathy and kindness toward other people
- Being respectful of differing opinions, viewpoints, and experiences
- Giving and gracefully accepting constructive feedback

**Reply by falkTX** [reply]
I dont see how it is hurtful at all. I am being honest above to the point where I state how such feature would happen.

What you ask for is simply unsupported.
So you either get your hands dirty and see about "fixing" it yourself or be okay with how things are.

**Reply by brylie** [reply]
I understand better that the feature doesn't exist and that it goes against the way Cardinal is architected. The honest appraisal was all needed to give a bit more insight.

I fully realize that my question may not have a solution. Similarly, I do not expect anyone to be obligated to write any code or reply related to my inquiry. However, I am hopeful that we can treat each other with empathy by respecting the diversity of abilities and perspectives when interacting online.

The hurtful part that may dissuade participation in the Cardinal community is the assumption that everyone who proposes an idea or question should "roll up their sleeves" by writing code. A healthy open-source community [acknowledges many types of contributions](https://allcontributors.org/), including ideas, discussions, support, and code, while acknowledging that not everyone shares the same knowledge or skillset.

**Reply by falkTX** [reply]
> A healthy open-source community [acknowledges many types of contributions](https://allcontributors.org/), including ideas, discussions, support, and code, while acknowledging that not everyone shares the same knowledge or skillset.

This doesnt work here though.
Cardinal follows what Rack does, mostly, as a way to remain compatible and also have less work in general for maintenance.

Discussions and support apply equality to Cardinal and Rack, since the core of how they work is basically the same.
But because Rack is not open to contributions, what can change in Cardinal in meaningful ways is always going to be very limited.

Again, I do not want to make this a big project, it was never the intention.
I already spend way too much time on Cardinal than I plan for, which will slow down considerably now that we kinda hit the limits of CI build time and resources.
tbqh I am against making this a "community" of sorts, it is just a hobby thing and I am fine keeping it that way. As a fun, interesting and learning project it has been very valuable.
If you want to have more than that, literally you do have to do some of that stuff yourself because me (the original Cardinal author) am not interested on pursuing that.

**Comment by johnallsup** [comment]
It's a quick hack. But all that is necessary is a text config file in a standard location. It just needs to be an include/exclude list, a bit like .gitignore. It needs a default (include all/exclude all). Then, include/exclude rules that match module names. And a means to dump a complete list of modules. Then the user can edit this text file. When the plugin loads, if it finds such a text config file, it filters the list of available modules from the browser. The sheer number of modules it has to filter through when using the search box is a pain, especially when 95% of the modules are ones I don't want to use for a given patch. When learning VCV/Cardinal, it would be good to have a small well chosen list. E.g. a dozen oscillators, half a dozen filters, a few envelopes and stuff. That's enough for a beginner to play with.

**Reply by dromer** [reply]
We have the CardinalMini for this. It's meant to be a very basic set of modules to do all the standard modular audio things.
(also work polyphonically and do not load external resources)

Otherwise: use favorites in the module browser!

**Reply by johnallsup** [reply]
CardinalMini: It only exists as LV2 and Standalone plugin, not VST3. So e.g. you can't use CardinalMini in most DAWs.

Second, the selection is fixed at compile time, not configurable afterwards with a simple text configuration. One issue with Cardinal is how laggy the browser is due to the sheer number of modules it's trying to search through, most of which a typical user won't need (which depends on the user, of course).

**Reply by dromer** [reply]
> CardinalMini: It only exists as LV2 and Standalone plugin, not VST3. So e.g. you can't use CardinalMini in most DAWs.

Oops, I forgot about this!
